core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java

```diff
@@ -42,12 +47,29 @@ void consumeIntermediateBestSolution(Solution_ bestSolution, BooleanSupplier isE
         }
     }
 
+    // Called on the Solver thread.
+    void consumeFirstInitializedSolution(Solution_ firstInitializedSolution) {
+        try {
+            // Called on the solver thread
+            // During the solving process, this lock is called once, and it won't block the Solver thread
+            firstSolutionConsumption.acquire();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException("Interrupted when waiting for the first initialized solution consumption.");
+        }
+        // called on the Consumer thread
+        this.firstInitializedSolution = firstInitializedSolution;
+        scheduleFirstInitializedSolutionConsumption();
+    }
+
     // Called on the Solver thread after Solver#solve() returns.
     void consumeFinalBestSolution(Solution_ finalBestSolution) {
         try {
             // Wait for the previous consumption to complete.
             // As the solver has already finished, holding the solver thread is not an issue.
             activeConsumption.acquire();
+            // Wait for the first solution consumption to complete
+            firstSolutionConsumption.acquire();
```

Comment:
Can the following (in theory) happen?

1. here, we acquire the `firstSolutionConsumption` lock
2. now if the init event comes, it waits on https://github.com/TimefoldAI/timefold-solver/pull/915/files#diff-07ca1817df09fb6ca99e48c7070b6e09d9992a9d31e7b02bd821bf62ecf52dedR55
3. the consumer finishes the final best solution and disposes the consumer thread
4. ` scheduleFirstInitializedSolutionConsumption` then fails due to the executor not accepting further tasks.

The important question is whether the final BSE can come before the init event and intuitively, I would say no.

Reply:
Yes, I agree with you and couldn't imagine a scenario where BSE could occur before the init event. The init event will always be triggered at the end of the custom or CH phase immediately before an LS phase. So, the `consumeFirstInitializedSolution` will acquire the lock before the LS executes, as only one phase will try to lock  `firstSolutionConsumption`.

Reply:
Devil's advocate here: the final CH (the same phase which triggers the FISE) also triggers a BSE. (FIS == BS most likely) Does this change anything?

Reply:
This situation is impossible. A final CH won't try to consume the FISE because it must immediately precede a LS, and its [flag](https://github.com/TimefoldAI/timefold-solver/pull/915/files#diff-a5c58876daf3e8f4f7fc44827a6c525a6a55e36d9491bbeb7a566b70c81317b4R305) wouldn't be enabled.